hngprojects/hng_boilerplate_nestjs
# feat: create organisation permissions guard

## What does this pull request do?
This pull request introduces a new guard, `OrganisationPermissionsGuard`, that checks if the user trying to access an endpoint has the required permissions for that organisation.

## How should this be manually tested?
1. Assign Roles and Permissions: Ensure that roles and permissions are assigned to users within an organization.
2. Access Protected Endpoints:
    - Try accessing endpoints protected by the `OrganisationPermissionsGuard` with different user roles.
    - Verify that users with appropriate permissions can access the endpoints.
    - Confirm that users without the necessary permissions receive the appropriate error response.
3. Test accessing the endpoint with different users:
   - User with `CanEditUsers` permission should successfully access the endpoint.
   - User without `CanEditUsers` permission should be denied access.
 
## How to create a new organisation permission and make use of the `OrganisationPermissionsGuard`
In this scenario, were going to create a permission called `CanGetProducts`
1.  Access `/src/modules/organisation-permissions/helpers/PermissionCategory.ts`. and append your new permission to the enum
![Screenshot 2024-08-07 222506](https://github.com/user-attachments/assets/904fbef6-ac0f-4aef-b5b5-31e56180eaf4)

3. Access `src/modules/organisation-permissions/dto/permission-list.dto.ts` and update it with the new permission you just created
![Screenshot 2024-08-07 224436](https://github.com/user-attachments/assets/b7a60468-f97f-4155-91ec-a7bbd0a7f165)

4. After following these steps, you can use the guard
![Screenshot 2024-08-07 224000](https://github.com/user-attachments/assets/be866bd8-3721-48ce-af6f-f7d15066ae66)

## Tasks done
- [x] Introduced OrganisationPermissionsGuard to validate user permissions.
- [x] Created Permissions decorator to set required permissions on routes.
- [x] Tested guard with various user roles and permissions.

Credits to @King-Mikaelson @dayve @MuthoniMN. They were of Immense help!
